NuRadioReco/modules/channelReadoutWindowCutter.py

```diff
@@ -0,0 +1,137 @@
+                        cut_samples_beginning = trigger_time_sample - samples_before_trigger
+                        roll_by = 0
+                        if(cut_samples_beginning + number_of_samples > trace_length):
+                            logger.warning("trigger time is sample {} but total trace length is only {} samples (requested trace length is {} with an offest of {} before trigger). To achieve desired configuration, trace will be rolled".format(
+                                trigger_time_sample, trace_length, number_of_samples, samples_before_trigger))
+                            roll_by = cut_samples_beginning + number_of_samples - trace_length  # roll_by is positive
+                            trace = np.roll(trace, -1 * roll_by)
+                            cut_samples_beginning -= roll_by
+                        rel_station_time_samples = cut_samples_beginning + roll_by
+                    elif(samples_before_trigger > trigger_time_sample):
+                        roll_by = -trigger_time_sample + samples_before_trigger
+                        logger.warning(f"trigger time is before 'trigger offset window' (requested samples before trigger = {samples_before_trigger}," \
+                                        f"trigger time sample = {trigger_time_sample}), the trace needs to be rolled by {roll_by} samples first" \
+                                        f" = {roll_by / sampling_rate/units.ns:.2f}ns")
+                        trace = np.roll(trace, roll_by)
+
+                    # shift trace to be in the correct location for cutting
+                    trace = trace[cut_samples_beginning:(number_of_samples + cut_samples_beginning)]
+                    channel.set_trace(trace, channel.get_sampling_rate())
+                    channel.set_trace_start_time(trigger_time - pre_trigger_time)
+                    # channel.set_trace_start_time(channel.get_trace_start_time() + rel_station_time_samples / channel.get_sampling_rate())
+                    # logger.debug(f"setting trace start time to {channel.get_trace_start_time() + rel_station_time_samples / channel.get_sampling_rate():.0f} = {channel.get_trace_start_time():.0f} + {rel_station_time_samples / channel.get_sampling_rate():.0f}")
+
+    
+
+    def __check_sampling_rates(self, detector_sampling_rate, channel_sampling_rate):
+        if not self.__sampling_rate_warning_issued: # we only issue this warning once
+            if not np.isclose(detector_sampling_rate, channel_sampling_rate):
+                logger.warning(
+                    'triggerTimeAdjuster was called, but the channel sampling rate '
+                    f'({channel_sampling_rate/units.GHz:.3f} GHz) is not equal to '
+                    f'the target detector sampling rate ({detector_sampling_rate/units.GHz:.3f} GHz). '
+                    'Traces may not have the correct trace length after resampling.'
+                )
+                self.__sampling_rate_warning_issued = True
```

Comment:
Do we need this warning at all? We ensure that we have the correct number of samples with:
```

                # this should ensure that 1) the number of samples is even and
                # 2) resampling to the detector sampling rate results in the correct number of samples
                # (note that 2) can only be guaranteed if the detector sampling rate is lower than the
                # current sampling rate)
                number_of_samples = int(
                    2 * np.ceil(
                        detector.get_number_of_samples(station.get_id(), channel.get_id()) / 2
                        * sampling_rate / detector_sampling_rate
                    ))
```
Maybe we can only check that `detector_sampling_rate <= channel_sampling_rate`?

Reply:
I think when I implemented this I thought about it sufficiently to realize that if `sampling_rate < detector_sampling_rate` we cannot guarantee the correct number of samples (with the way resampling is currently implemented); I am not actually sure if the correct number of samples is guaranteed in all cases if `sampling_rate > detector_sampling_rate`; my feeling was that this is only true in most cases which is why I left the warning as it is. If you can work out the maths and figure out that this is wrong and we don't need the warning we can change this : )